### PR TITLE
[SPARK-23372][SQL] Writing empty struct in parquet fails during execution. It should fail earlier in the processing.

### PR DESCRIPTION
**Pull request by dilipbiswal**
## What changes were proposed in this pull request?
Currently we allow writing data frames with empty schema into a file based datasource for certain file formats such as JSON, ORC etc. For formats such as Parquet and Text, we raise error at different times of execution. For text format, we return error from the driver early on in processing where as for format such as parquet, the error is raised from executor.

**Example**
spark.emptyDataFrame.write.format("parquet").mode("overwrite").save(path)
**Results in**
``` SQL
org.apache.parquet.schema.InvalidSchemaException: Cannot write a schema with an empty group: message spark_schema {
 }

at org.apache.parquet.schema.TypeUtil$1.visit(TypeUtil.java:27)
 at org.apache.parquet.schema.TypeUtil$1.visit(TypeUtil.java:37)
 at org.apache.parquet.schema.MessageType.accept(MessageType.java:58)
 at org.apache.parquet.schema.TypeUtil.checkValidWriteSchema(TypeUtil.java:23)
 at org.apache.parquet.hadoop.ParquetFileWriter.<init>(ParquetFileWriter.java:225)
 at org.apache.parquet.hadoop.ParquetOutputFormat.getRecordWriter(ParquetOutputFormat.java:342)
 at org.apache.parquet.hadoop.ParquetOutputFormat.getRecordWriter(ParquetOutputFormat.java:302)
 at org.apache.spark.sql.execution.datasources.parquet.ParquetOutputWriter.<init>(ParquetOutputWriter.scala:37)
 at org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat$$anon$1.newInstance(ParquetFileFormat.scala:151)
 at org.apache.spark.sql.execution.datasources.FileFormatWriter$SingleDirectoryWriteTask.newOutputWriter(FileFormatWriter.scala:376)
 at org.apache.spark.sql.execution.datasources.FileFormatWriter$SingleDirectoryWriteTask.execute(FileFormatWriter.scala:387)
 at org.apache.spark.sql.execution.datasources.FileFormatWriter$$anonfun$org$apache$spark$sql$execution$datasources$FileFormatWriter$$executeTask$3.apply(FileFormatWriter.scala:278)
 at org.apache.spark.sql.execution.datasources.FileFormatWriter$$anonfun$org$apache$spark$sql$execution$datasources$FileFormatWriter$$executeTask$3.apply(FileFormatWriter.scala:276)
 at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1411)
 at org.apache.spark.sql.execution.datasources.FileFormatWriter$.org$apache$spark$sql$execution$datasources$FileFormatWriter$$executeTask(FileFormatWriter.scala:281)
 at org.apache.spark.sql.execution.datasources.FileFormatWriter$$anonfun$write$1.apply(FileFormatWriter.scala:206)
 at org.apache.spark.sql.execution.datasources.FileFormatWriter$$anonfun$write$1.apply(FileFormatWriter.scala:205)
 at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
 at org.apache.spark.scheduler.Task.run(Task.scala:109)
 at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:345)
 at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
 at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
 at java.lang.Thread.run(Thread.
```

In this PR, we unify the error processing and raise error on attempt to write empty schema based dataframes into file based datasource (orc, parquet, text , csv, json etc) early on in the processing.

## How was this patch tested?

Unit tests added in FileBasedDatasourceSuite.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala [hunk 1]
```diff
@@ -68,6 +68,16 @@ class ParquetFileFormat
 
   override def toString: String = "Parquet"
 
+  private def verifySchema(schema: StructType): Unit = {
+    if (schema.size < 1) {
```
Comment (mannharleen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: any particular reason for using "< 1" and not "== 0". just curious

Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: no particular reason :-). I can change that to "== 0".

File: sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala [hunk 2]
```diff
@@ -72,6 +72,29 @@ class FileBasedDataSourceSuite extends QueryTest with SharedSQLContext {
...
+    test(s"SPARK-23372 writing empty dataframe should produce AnalysisException - $format") {
+      withTempPath { outputPath =>
+        intercept[AnalysisException] {
+          spark.emptyDataFrame.write.format(format).save(outputPath.toString)
+        }
+      }
+    }
+  }
+
+  // Formats excluding text and parquet allow writing empty data frames to files.
```
Comment (dilipbiswal) [inline sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala]: I have added this test to show the current behaviour of reading the empty data frame.  For formats like orc, json, csv , we succeed to write an empty data frame but get the error while reading while inferring the schema ? Is this the right behaviour ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala [hunk 3]
```diff
@@ -68,6 +68,16 @@ class ParquetFileFormat
 
   override def toString: String = "Parquet"
 
+  private def verifySchema(schema: StructType): Unit = {
+    if (schema.size == 0) {
+      throw new AnalysisException(
+        s"""
+           |Parquet data source does not support writing empty groups.
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: `group` is a parquet term. Let's use `schema` instead?

We should also check for nested empty schema's.

File: sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala [hunk 4]
```diff
@@ -72,6 +72,29 @@ class FileBasedDataSourceSuite extends QueryTest with SharedSQLContext {
...
+  allFileBasedDataSources.filterNot(p => p == "text" || p == "parquet").foreach { format =>
+    test(s"SPARK-23372 writing empty dataframe and reading from it - $format") {
+      withTempPath { outputPath =>
+          spark.emptyDataFrame.write.format(format).save(outputPath.toString)
+          intercept[AnalysisException] {
+            val df = spark.read.format(format).load(outputPath.toString)
```
Comment (hvanhovell) [inline sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala]: This should pass right?

**Comment by hvanhovell** [comment]
@dilipbiswal this is a nice improvement. I left a few comments.

File: sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala (hunk 4 above)
Comment (dilipbiswal) [reply sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala]: @hvanhovell Actually thats my question as well. Please see my comment [comment/question](https://github.com/apache/spark/pull/20579#issuecomment-364994881)

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala [hunk 5]
```diff
@@ -72,6 +72,29 @@ class FileBasedDataSourceSuite extends QueryTest with SharedSQLContext {
     }
   }
 
+  // Text and Parquet format does not allow wrting data frame with empty schema.
+  Seq("parquet", "text").foreach { format =>
+    test(s"SPARK-23372 writing empty dataframe should produce AnalysisException - $format") {
+      withTempPath { outputPath =>
+        intercept[AnalysisException] {
+          spark.emptyDataFrame.write.format(format).save(outputPath.toString)
+        }
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala]: Can we check the error message to make it sure?

Comment (dilipbiswal) [reply sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala]: Sure @dongjoon-hyun

File: sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala (hunk 4 above)
Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala]: Sorry if I misunderstood. The link is https://github.com/apache/spark/pull/20579#issuecomment-364994881. Is that the right link?

**Comment by cloud-fan** [comment]
I think this should be applied to all data sources not only parquet. I can't think of any cases that a data source needs to write data with empty schema, cc @rdblue for confirmation.

**Comment by rdblue** [comment]
I agree, we should probably add a check for storing a DataFrame with no columns for now. This is normally caught by the pre-insert rules, but since the table is getting "created" in this case there is nothing to check. 

In the future, I think that the create and insert should be logically separate so that the create will fail and complain that you can't create a table without at least one column. I think it will be cleaner to separate concerns like that.

**Comment by dilipbiswal** [comment]
@cloud-fan @rdblue Thank you for clarification. I am sorry, i hadn't seen your comments before i pushed the last change which targets only parquet. I will adjust the fix to target all formats in a future commit. Thanks again !!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 6]
```diff
@@ -542,6 +542,11 @@ case class DataSource(
       throw new AnalysisException("Cannot save interval data type into external storage.")
     }
 
+    if (data.schema.size == 0) {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: Is it required? This is a behavior change. Can we exclude it from this PR?

Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: @gatorsmile May i request you to please quickly go through Wenchen's and Ryan's comments  above ? My understanding is that , we want to consistently rejecting writing empty schema for all the data sources ? Please let me know.

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: Currently, we are not blocking this. I do not think we should introduce this behavior change. This is risky to block all the cases. 

Previously, I tried to block CREATE TABLE with an empty schema. Later, I hit a regression because some data sources are using options/table properties to specify the schema...

A general guide here is to avoid behavior changes if possible. When we have to introduce a behavior change, we should make it configurable. At least, users can convert it back by using a flag.

Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: @gatorsmile @cloud-fan OK.. sounds reasonable to me. I will rollback the latest change in this PR and we can discuss if we want to introduce the behaviour change in a future jira/pr.

**Comment by cloud-fan** [comment]
After some more thoughts, we should try our best to not introduce behavior change to existing data sources. How about we only add this check for file-based data sources(all of them are built-in)? We can have a follow-up JIRA to add this check for data source v2.

**Comment by dilipbiswal** [comment]
@cloud-fan OK.. i was thinking of adding this check in each built in datasource like Text, CSV, Parquet, ORC, JSON etc. Just like we check it in Parquet with this PR.  Would you have any concern with that approach ? That would some duplicate code under each specific format , but gives us the flexibility to change the behaviour for a datasource should we need ? What do you think ?

**Comment by gatorsmile** [comment]
Check whether the format is file-based data sources? Then, we do not need to check the same thing for different file sources.

**Comment by dilipbiswal** [comment]
@gatorsmile Thank you Sean. I will follow your suggestion.

**Comment by dilipbiswal** [comment]
retest this please

**Comment by dilipbiswal** [comment]
@gatorsmile When you get a chance, could you please see if the check for internal datasource looks reasonable ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 7]
```diff
@@ -546,6 +546,10 @@ case class DataSource(
       case dataSource: CreatableRelationProvider =>
         SaveIntoDataSourceCommand(data, dataSource, caseInsensitiveOptions, mode)
       case format: FileFormat =>
+        if (DataSource.isBuiltInFileBasedDataSource(format) && data.schema.size == 0) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: We don't need this check. `FileFormat` is internal, we don't need to distinguish between built-in file format or external ones.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: actually we just need [this check](https://github.com/apache/spark/pull/20579/files#diff-ee26d4c4be21e92e92a02e9f16dbc285R71) here.

**Comment by dilipbiswal** [comment]
@cloud-fan Thank you. I assumed (wrongly) that we don't want to change the behaviour for an external file based datasource and we wanted to scope our check only to spark built in data sources. I have made the change based on your suggestion. I have parked the verifySchema method in DataSource for now. Pl. let me know if thats the right place or we want to move it to a Utility class ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 8]
```diff
@@ -719,4 +720,27 @@ object DataSource extends Logging {
...
     globPath
   }
+
+  /**
+   * Called before writing into a FileFormat based data source to make sure the
+   * supplied schema is not empty.
+   * @param schema
+   */
+  private def verifySchema(schema: StructType): Unit = {
+    def verifyInternal(schema: StructType): Boolean = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: better to call it `hasEmptySchema`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala [hunk 9]
```diff
@@ -77,7 +77,6 @@ class ParquetFileFormat
       job: Job,
       options: Map[String, String],
       dataSchema: StructType): OutputWriterFactory = {
-
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: unnecessary change

**Comment by dilipbiswal** [comment]
@cloud-fan ok.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala (hunk 9 above)
Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: @cloud-fan will remove.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala (hunk 8 above)
Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: @cloud-fan will make the change.

**Comment by dilipbiswal** [comment]
retest this please

**Comment by dilipbiswal** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 10]
```diff
@@ -719,4 +720,27 @@ object DataSource extends Logging {
...
     globPath
   }
+
+  /**
+   * Called before writing into a FileFormat based data source to make sure the
+   * supplied schema is not empty.
+   * @param schema
+   */
+  private def hasEmptySchema(schema: StructType): Unit = {
+    def hasEmptySchemaInternal(schema: StructType): Boolean = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: they should be `verifySchema` and `hasEmptySchema`, depending on their return type.

Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: You are right @cloud-fan. Given we are raising the error from the function itself, should i rename it to validateSchema ?

Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: @cloud-fan I have gone ahead and changed the top level function name to validateSchema. I have kept the internal function name to be hasEmptySchema. Hopefully it makes sense now.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by dilipbiswal** [comment]
Thanks a lot @cloud-fan @gatorsmile
